providers/dns/nicru/nicru.go

```diff
@@ -0,0 +1,329 @@
+package nicru
+
+import (
+	"bytes"
+	"encoding/json"
+	"encoding/xml"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/xenolf/lego/acme"
+)
+
+var baseURL = "https://api.nic.ru"
+
+// DNSProvider is an implementation of the acme.ChallengeProvider interface that uses
+// DNSMadeEasy's DNS API to manage TXT records for a domain.
+type DNSProvider struct {
+	baseURL     string
+	apiUsername string
+	apiPassword string
+	apiID       string
+	apiSecret   string
+	serviceName string
+	auth        *authResponse
+	client      http.Client
+}
+
+type authResponse struct {
+	ExpiresIn   string `json:"expires_in"`
+	AccessToken string `json:"access_token"`
+	TokenType   string `json:"token_type"`
+}
+
+type authError struct {
+	Error string `json:"error"`
+}
+
+// Record holds the nic.ru API representation of a DNS Record
+type Record struct {
+	ID       int    `json:"id"`
+	Type     string `json:"type"`
+	Name     string `json:"name"`
+	Value    string `json:"value"`
+	TTL      int    `json:"ttl"`
+	SourceID int    `json:"sourceId"`
+}
+
+// NewDNSProvider returns a DNSProvider instance configured for nic.ru DNS
+func NewDNSProvider() (*DNSProvider, error) {
+	nicruUsername := os.Getenv("NICRU_USERNAME")
+	nicruPassword := os.Getenv("NICRU_PASSWORD")
+	nicruAPIKey := os.Getenv("NICRU_API_KEY")
+	nicruAPISecret := os.Getenv("NICRU_API_SECRET")
+	nicruServiceName := os.Getenv("NICRU_SERVICE_NAME")
+	provider := &DNSProvider{
+		baseURL: baseURL,
+
+		apiUsername: nicruUsername,
+		apiPassword: nicruPassword,
+		apiID:       nicruAPIKey,
+		apiSecret:   nicruAPISecret,
+		serviceName: nicruServiceName,
+
+		client: http.Client{Timeout: 30 * time.Second},
+	}
+
+	err := provider.getAuthRespone()
+	if err != nil {
+		return nil, err
+	}
+	return provider, nil
+}
+
+// getCredentials uses the supplied credentials to return a
+// DNSProvider instance configured for nic.ru
+func (d *DNSProvider) getAuthRespone() error {
+	// TODO: validate all credential are set
```

Comment:
Should we wait to merge this until this is done? I imagine if credentials aren't set, the requests fail anyway.